test/cpp/api/dataloader.cpp

```diff
@@ -833,57 +834,83 @@ TEST(DataLoaderTest, RespectsTimeout) {
   ASSERT_LT(duration.count(), 1);
 }
 
+struct Barrier {
+  void wait(size_t target) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    ++counter_;
+    cv_.wait(lock, [this, target] { return this->counter_ == target; });
+    cv_.notify_all();
+  }
+
+  size_t counter_{0};
+  std::condition_variable cv_;
+  std::mutex mutex_;
+};
+
+namespace ordering_test {
 namespace {
-std::atomic<size_t> ordering_test_counter{0};
-std::condition_variable ordering_test_cv;
-std::mutex ordering_test_mutex;
-const std::array<size_t, 4> ordering_test_order = {3, 1, 0, 2};
-std::atomic<size_t> ordering_test_index{0};
+const size_t kNumberOfWorkers = 10;
+const std::vector<size_t> kOrderInWhichWorkersReturnTheirBatch =
+    {3, 7, 0, 5, 4, 8, 2, 1, 9, 6};
 } // namespace
 
-struct OrderingTestDataset : datasets::BatchDataset<DummyDataset, int> {
-  OrderingTestDataset() = default;
+struct Dataset : datasets::BatchDataset<Dataset, size_t> {
+  Dataset() = default;
 
   // This copy constructor will be called when we copy the dataset into a
   // particular thread.
-  OrderingTestDataset(const OrderingTestDataset& other)
-      : id(ordering_test_counter++) {}
-
-  OrderingTestDataset(OrderingTestDataset&& other) noexcept = default;
-  OrderingTestDataset& operator=(const OrderingTestDataset& other) = delete;
-  OrderingTestDataset& operator=(OrderingTestDataset&& other) noexcept = delete;
-
-  int get_batch(torch::ArrayRef<size_t> indices) override {
-    std::unique_lock<std::mutex> lock(ordering_test_mutex);
-    // block until order.at(index) == my_thread_id (until it's this thread's
-    // turn)
-    ordering_test_cv.wait(lock, [this] {
-      return ordering_test_order.at(ordering_test_index.load()) == this->id;
-    });
-    // Make one step in the order.
-    ++ordering_test_index;
+  Dataset(const Dataset& other) {
+    static std::atomic<size_t> counter{0};
+    thread_id_ = counter.fetch_add(1);
+  }
+
+  Dataset(Dataset&& other) noexcept = default;
+  Dataset& operator=(const Dataset& other) = delete;
+  Dataset& operator=(Dataset&& other) noexcept = delete;
+
+  size_t get_batch(torch::ArrayRef<size_t> indices) override {
+    static Barrier barrier;
+    static auto order_iterator = kOrderInWhichWorkersReturnTheirBatch.begin();
+    static std::condition_variable cv;
+    static std::mutex mutex;
+
+    // Wait for all threads to get an index batch and arrive here.
+    barrier.wait(kNumberOfWorkers);
```

Comment:
OK, so if I understand correctly, the old code could have let a thread out before all the indices were assigned to worker thread, and so after it (very speedily) finished processing its batch, it would grab another one, starving the actual thread. In that case, I would expect the code to have deadlocked, because now the thread is never going to see its index show up again in the test order.

You've solved this by forcing all `get_batch` calls to block until the worker threads have each grabbed a Dataset. But... I kind of don't see why, fundamentally, `enforce_ordering` shouldn't work even if a worker thread finishes up their work and starts working on another index. You're still releasing threads based on their thread ID, but shouldn't you release them on the value of `indices.front()`? That's the thing you actually care about!

Reply:
The first paragraph is correct. About the second paragraph: your observations are true, and the reason I do this is purely to simulate a deterministic *but unsorted* order in which threads are unblocked, to test that the ordering really takes place (otherwise maybe threads wold return in order, in which case the ordering functionality wouldn't really be tested). IRL threads could fetch multiple batches and that would be fine